xdist/scheduler/loadsuite.py

```diff
@@ -0,0 +1,413 @@
+            return False
+
+        for assigned_suites in self.assigned_work.values():
+            if self._pending_of(assigned_suites) >= 2:
+                return False
+
+        return True
+
+    @property
+    def has_pending(self):
+        """
+        Return True if there are pending test items.
+
+        This indicates that collection has finished and nodes are still
+        processing test items, so this can be thought of as
+        "the scheduler is active".
+        """
+        if self.workqueue:
+            return True
+
+        for assigned_suites in self.assigned_work.values():
+            if self._pending_of(assigned_suites) > 0:
+                return True
+
+        return False
+
+    def add_node(self, node):
+        """
+        Add a new node to the scheduler.
+
+        From now on the node will be assigned work units to be executed.
+
+        Called by the ``DSession.slave_slaveready`` hook when it successfully
+        bootstraps a new node.
+        """
+        assert node not in self.assigned_work
+        self.assigned_work[node] = OrderedDict()
+
+    def remove_node(self, node):
+        """
+        Remove a node from the scheduler.
+
+        This should be called either when the node crashed or at shutdown time.
+        In the former case any pending items assigned to the node will be
+        re-scheduled.
+
+        Called by the hooks:
+
+        - ``DSession.slave_slavefinished``.
+        - ``DSession.slave_errordown``.
+
+        Return the item being executed while the node crashed or None if the
+        node has no more pending items.
+        """
+        workload = self.assigned_work.pop(node)
+        if not self._pending_of(workload):
+            return None
+
+        # The node crashed, identify test that crashed
+        for work_unit in workload.values():
+            for nodeid, completed in work_unit.items():
+                if not completed:
+                    crashitem = nodeid
+                    break
+            else:
+                continue
+            break
+        else:
+            raise RuntimeError(
+                'Unable to identify crashitem on a workload with '
+                'pending items'
+            )
+
+        # Made uncompleted work unit available again
+        self.workqueue.update(workload)
+
+        for node in self.assigned_work:
+            self._reschedule(node)
+
+        return crashitem
+
+    def add_node_collection(self, node, collection):
+        """
+        Add the collected test items from a node.
+
+        The collection is stored in the ``.registered_collections`` dictionary.
+
+        Called by the hook:
+
+        - ``DSession.slave_collectionfinish``.
+        """
+        assert node in self.assigned_work
+
+        if not self.collection_is_completed:
+            self.registered_collections[node] = list(collection)
+            return
+
+        # A new node has been added later, perhaps an original one died.
```

Comment:
When this happens, shouldn't we register the new node and its collection, like it is done in line `225` above?